pkg/volume/rbd/rbd.go

```diff
@@ -357,6 +360,131 @@ func (plugin *rbdPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*vol
 	return volume.NewSpecFromVolume(rbdVolume), nil
 }
 
+// <><><>
+
+func (plugin *rbdPlugin) ConstructBlockVolumeSpec(podUID types.UID, volumeName, mapPath string) (*volume.Spec, error) {
+	pluginDir := plugin.host.GetVolumeDevicePluginDir(rbdPluginName)
+	blkutil := volutil.NewBlockVolumePathHandler()
+
+	globalMapPathUUID, err := blkutil.FindGlobalMapPathUUIDFromPod(pluginDir, mapPath, podUID)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(5).Infof("globalMapPathUUID: %v, err: %v", globalMapPathUUID, err)
+	globalMapPath := filepath.Dir(globalMapPathUUID)
+	if len(globalMapPath) == 1 {
+		return nil, fmt.Errorf("failed to retreive volume plugin information from globalMapPathUUID: %v", globalMapPathUUID)
+	}
+	return getVolumeSpecFromGlobalMapPath(volumeName, globalMapPath)
+}
+
+func getVolumeSpecFromGlobalMapPath(volumeName, globalMapPath string) (*volume.Spec, error) {
+	// Retreive volume spec information from globalMapPath
+	// globalMapPath example:
+	//   plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumePluginDependentPath}
+	rbdVolume := &v1.Volume{
```

Comment:
I guess current ConstructVolumeSpec() doesn't create enough volumeSpec.

At reconstructVolume(), volumeSpec is used to create mounter or blockVolumeMapper based on volumeMode. VolumeSpec should have enough information to pass one of these methods.
- NewMounter(): Filesystem volume case
- NewBlockVolumeMapper(): Block volume case

Can you check if volume reconstruction works properly?

Reply:
The volume reconstruction happens when volume is deleted during kubelet node is down.
We can test this situation by
- Start cluster with local-up-cluster.sh
- Create volume
- Ctrl+C for local-up-cluster.sh
- Start cluster with local-up-cluster.sh again

After second local-up-cluster.sh, the volume is still attached and pod volume dir, plugin dir are still exists on kubelet node. Therefore, kubelet tries to reconstruct volume from these information.
You can see/debug what happens during second cluster up.